File: mojo/edk/system/android/parcelable_channel_client.cc
// Copyright 2017 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "mojo/edk/system/android/parcelable_channel_client.h"

#include "mojo/edk/system/android/jni/ParcelableChannelClient_jni.h"

namespace mojo {
namespace edk {

// static
ParcelableChannelClient ParcelableChannelClient::CreateProxyChannelClient() {}

// static
ParcelableChannelClient ParcelableChannelClientCreateChannelClient(
    const base::android::JavaRef<jobject>& iparcelable_channel) {}

ParcelableChannelClient::ParcelableChannelClient() {}

ParcelableChannelClient::~ParcelableChannelClient() {}

ParcelableChannelClient::ParcelableChannelClient(
    const base::android::JavaRef<jobject>& j_parcelable_channel)
    : j_parcelable_channel_(j_parcelable_channel) {
  DCHECK(!j_parcelable_channel_.is_null());
}

ParcelableChannelClient::ParcelableChannelClient(
    ParcelableChannelClient&& other) = default;

ParcelableChannelClient& ParcelableChannelClient::operator=(
    ParcelableChannelClient&& other) = default;

void ParcelableChannelClient::SendParcelable(
    uint32_t id,
    base::android::ScopedJavaLocalRef<jobject> parcelable) {
  DCHECK(!j_parcelable_channel_.is_null());

  Java_ParcelableChannelClient_sendParcelable(
      base::android::AttachCurrentThread(), j_parcelable_channel_, id,
      parcelable);
}

}  // namespace edk
}  // namespace mojo

File: mojo/edk/system/android/java/src/org/chromium/mojo/edk/ParcelableChannelClient.java
// Copyright 2017 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.mojo.edk;

import android.os.IBinder;
import android.os.Parcelable;
import android.os.RemoteException;

import org.chromium.base.Log;
import org.chromium.base.annotations.CalledByNative;
import org.chromium.base.annotations.JNINamespace;

/**
 * Wraps a IParcelableChannel so it can be called from native code.
 */
@JNINamespace("mojo::edk")
public final class ParcelableChannelClient {
    private static final String TAG = "ParcelChanClient";

    private IParcelableChannel mIParcelableChannel;

    private ParcelableChannelClient(IParcelableChannel parcelableChannel) {
        mIParcelableChannel = parcelableChannel;
    }

    @CalledByNative
    public static ParcelableChannelClient create(IBinder parcelableChannel) {
        return new ParcelableChannelClient(IParcelableChannel.Stub.asInterface(parcelableChannel));
    }

    @CalledByNative
    public boolean sendParcelable(int id, Parcelable parcelable) {
        try {
            mIParcelableChannel.sendParcelable(new ParcelableWrapper(id, parcelable));
            return true;
        } catch (RemoteException e) {
            Log.e(TAG, "Failed to send parcelable.", e);
            return false;
        }
    }
}